lib/Parse/SyntaxParsingContext.cpp

```diff
@@ -157,42 +166,49 @@ const SyntaxParsingContext *SyntaxParsingContext::getRoot() const {
   return Curr;
 }
 
+ParsedTokenSyntax SyntaxParsingContext::popToken() {
+  return popIf<ParsedTokenSyntax>().getValue();
+}
+
 /// Add Token with Trivia to the parts.
 void SyntaxParsingContext::addToken(Token &Tok, Trivia &LeadingTrivia,
                                     Trivia &TrailingTrivia) {
   if (!Enabled)
     return;
 
-  auto Arena = getArena();
-  auto Text = OwnedString::makeRefCounted(Tok.getText());
-
-  addRawSyntax(getTokenCache().getToken(
-      Arena, Tok.getKind(), Text, LeadingTrivia.Pieces, TrailingTrivia.Pieces));
+  ParsedRawSyntaxNode raw;
+  if (IsBacktracking)
+    raw = ParsedRawSyntaxNode::makeDeferred(Tok, LeadingTrivia, TrailingTrivia);
```

Comment:
IIRC, current `Lexer` creates `TriviaPiece` with refcounted owned text. It would be nice if `Lexer` creates source referencing `TriviaPiece`s, then `SyntaxTreeCreator::recordToken()` turns them into "owned".

Reply:
I agree, and actually my intention is to make changes for the `Lexer` to only determine the source ranges for leading and trailing trivia, and only do the processing to determine the actual trivia pieces at `record` time, in `ParsedRawSyntaxRecorder`. That will prevent re-processing the same trivia pieces when doing backtracking.

However, I'd like to hold off on performance optimizations until after SwiftSyntax adopts the new model.